libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -2,41 +2,121 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import { Injectable } from '@nestjs/common';
+import { EligibilityService } from '@fxa/payments/eligibility';
+import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
 import {
   StripeClient,
   StripeManager,
   StripeSubscription,
+  SubplatInterval,
+  TaxAddress,
 } from '@fxa/payments/stripe';
-import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
-import { Injectable } from '@nestjs/common';
-import { ResultCart } from './cart.types';
 import { CheckoutError, CheckoutPaymentError } from './checkout.error';
+import {
+  CartTotalMismatchError,
+  CartEligibilityMismatchError,
+  CartEmailNotFoundError,
+} from './cart.error';
+import { CartManager } from './cart.manager';
+import { ResultCart } from './cart.types';
+import { handleEligibilityStatusMap } from './cart.utils';
+import { ContentfulService } from '@fxa/shared/contentful';
+import { AccountManager } from '@fxa/shared/account/account';
+import { DEFAULT_LOCALE } from '@fxa/shared/l10n';
 
 @Injectable()
 export class CheckoutService {
   constructor(
     private stripeClient: StripeClient,
     private stripeManager: StripeManager,
     private paypalCustomerManager: PaypalCustomerManager,
-    private paypalManager: PayPalManager
+    private paypalManager: PayPalManager,
+    private cartManager: CartManager,
+    private eligibilityService: EligibilityService,
+    private contentfulService: ContentfulService,
+    private accountManager: AccountManager
   ) {}
 
   async prePaySteps(cart: ResultCart) {
-    // TODO:
-    // - If uid not present, create user stub account and update cart
-    // - Check if customer exists, create via stripe manager and update cart
-    // - Make sure shipping address of customer is updated with cart taxAddress
-    // - Check if customer is eligible for product in cart
-    // - Check if amount to be charged matches amount in cart
-    // - Cancel incomplete subscriptions to price user is subscribing to
-
-    const { stripeCustomerId } = cart;
-    if (!stripeCustomerId) {
-      // TODO: this is a placeholder and won't be necessary once the above todo items are done
-      throw new Error("Stripe customer doesn't exist");
+    let customer;
+    const taxAddress = cart.taxAddress as any as TaxAddress;
+
+    if (!cart.email) {
+      throw new CartEmailNotFoundError(cart.id);
+    }
+
+    // if uid not found, create account stub and update cart
+    if (!cart.uid) {
+      const uid = await this.accountManager.createAccountStub(
+        cart.email,
+        cart.version,
+        DEFAULT_LOCALE
```

Comment:
To clarify - so I should add it as a required parameter then and update where applicable, as it does not exist right now?

Reply:
Yup that's it. 👍

Reply:
As per Slack, I didn't do it as a required parameter because that seems rather odd (that is - customer gets returned and should already have that attached; the stub account is only created if the customer does not exist). I instead use the locale for the stripe customer and if that doesn't exist then the fallback is the default locale.

Reply:
Unfortunately, the Stripe Customers `preferred_locales` isn't automatically populated, and would be an empty array if not explicitly set. So effectively this would always default to the `DEFAULT_LOCALE`.

I'd recommend taking a similar approach to the existing SP2.5 logic, which reads the `accept-language` request header to get the locale. 

In SP3.0 reading the users locale is currently done in Payments Next, which is why I initially recommended adding a required parameter to `prePaySteps` so that it can be passed in from the Payment Next, when it calls the Server Action.

Reply:
Since `prePaySteps` doesn't exist in 2.5, where would I find this?

Reply:
The similar approach to SP2.5 is getting the users locale from the `request`, `accept-language` header. This is currently done directly in the `AccountHandler.accountStub` method in `fxa-auth-server`.

However, in SP3.0 you can't do it directly in `prePaySteps`, since it doesn't have access to the `request`. But, Payments Next does have access to the request, and already has logic to get the users locale from the `request` `accept-language` header. 

So adding a parameter to `prePaySteps` and the Server Action, (and all methods/function in between), to get the user locale is probably the simplest way to go.

I hope that clarifies things a bit more than my initial comment. If not, I'm happy to discuss on Slack.

Reply:
Immensely! Thank you! :mozheart: